### PR TITLE
Completed model for god class detection

### PR DESCRIPTION
**Pull request by katarina-glorija**
Code refactored according to suggestions. God class detection implemented and tested.

### CODE REVIEW

File: SmellDetector/Detectors/RuleEngines/GodClassRuleNMDandNAD.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Collections.Generic;
+using SmellDetector.Controllers;
+using SmellDetector.SmellModel;
+using SmellDetector.SmellModel.Reports;
+
+namespace SmellDetector.Detectors.RuleEngines
+{
+    internal class GodClassRuleNMDandNAD : IDetector
```
Comment (Luburic) [inline SmellDetector/Detectors/RuleEngines/GodClassRuleNMDandNAD.cs:9]: Do we need this now that we have ClassMetricRE?

File: SmellDetector/Detectors/RuleEngines/MetricCriteria.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
+﻿using System;
+using System.Collections.Generic;
+using SmellDetector.Controllers;
+
+namespace SmellDetector.Detectors.RuleEngines
+{
+    public class MetricCriteria : Criteria
+    {
+        //This should be a single metric, so that we can compute the rule.
```
Comment (Luburic) [inline SmellDetector/Detectors/RuleEngines/MetricCriteria.cs:9]: Avoid comments - they are evil https://youtu.be/HskIvd9bpik

File: SmellDetector/Detectors/RuleEngines/Rule.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        private readonly Criteria _criteria;
+        private readonly SmellType _smellType;
+        public Rule(String DOI, Criteria criteria, SmellType smellType)
+        {
+            _DOI = DOI;
+            _criteria = criteria;
+            _smellType = smellType;
+        }
+
+        public Issue IsTriggered(String codeSnippetId, Dictionary<String, double> metrics)
```
Comment (Luburic) [inline SmellDetector/Detectors/RuleEngines/Rule.cs:20]: The "IsTriggered" name signals that a boolean should be returned. "Validate" might be more suitable.

File: SmellDetector/SmellModel/CaDETClassDTO.cs [hunk 4]
```diff
@@ -1,16 +1,22 @@
...
+﻿using System;
+using System.Collections.Generic;
 
 namespace SmellDetector.SmellModel
 {
     public class CaDETClassDTO
     {
-        public Dictionary<string, MetricsDTO> CodeItemMetrics { get; set; }
+        public string FullName { get; set; }
+        public Dictionary<string, double> CodeItemMetrics { get; set; }
```
Comment (Luburic) [inline SmellDetector/SmellModel/CaDETClassDTO.cs:9]: CodeItem is a deprecated term we used last year. CodeSnippet is what we use now, so refactor all occurences.

File: SmellDetector/SmellModel/CodeSnippetCollectionDTO.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace SmellDetector.SmellModel
+{
+    public class CodeSnippetCollectionDTO
+    {
+        public Dictionary<string, MetricsDTO> CodeItemMetrics { get; set; }
+
+        public CodeSnippetCollectionDTO()
+        {
+            CodeItemMetrics = new Dictionary<string, MetricsDTO>();
```
Comment (Luburic) [inline SmellDetector/SmellModel/CodeSnippetCollectionDTO.cs:12]: CodeSnippetMetrics.

**Review by Luburic — CHANGES_REQUESTED** [review]
Minor improvements inline.

File: SmellDetector/Detectors/RuleEngines/MetricCriteria.cs (hunk 2 above)
Comment (katarina-glorija) [reply SmellDetector/Detectors/RuleEngines/MetricCriteria.cs]: I forgot to delete this comment. I wrote it so I don't forget what's the purpose for this field.

File: SmellDetector/Detectors/RuleEngines/MetricCriteria.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        private readonly OperationEnum _operation;
+        private readonly double _threshold;
+        public MetricCriteria(String metric, OperationEnum operation, double threshold)
+        {
+            _metric = metric;
+            _operation = operation;
+            _threshold = threshold;
+        }
+
+        //This method should calculate the rule based on metric, operator and threshold and return true if the rule applies, otherwise false.
```
[the remark this answers is not in the document]
Comment (katarina-glorija) [reply SmellDetector/Detectors/RuleEngines/MetricCriteria.cs]: I like to write comments to remember what I want to implement, but forget to delete them when I'm done.

**Review by Luburic — COMMENTED** [review]
Before merging the changes into master, make sure you pull new commits from master into your branch and resolve any conflicts. When you do this and perform any additional cleanup you had in mind, assign me as reviewer so we can complete this PR.

**Pull request merged**